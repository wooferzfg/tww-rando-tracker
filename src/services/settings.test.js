import _ from 'lodash';

import FLAGS from '../data/flags.json';

import Permalink from './permalink';
import Settings from './settings';

describe('Settings', () => {
  beforeEach(() => {
    Settings.reset();
  });

  describe('initializeFromPermalink', () => {
    describe('default settings', () => {
      beforeEach(() => {
        Settings.initializeFromPermalink('MS4xMC4wAHllZXQABwEDAAygvgMA0AACAAAAAAGAIAAA');
      });

      test('initializes the version', () => {
        expect(Settings.version).toEqual('1.10.0');
      });

      test('initializes the starting gear', () => {
        expect(Settings.startingGear).toMatchSnapshot();
      });

      test('initializes the options', () => {
        expect(Settings.options).toMatchSnapshot();
      });

      test('initializes the flags', () => {
        expect(Settings.flags).toEqual([
          Settings.FLAGS.DUNGEON,
          Settings.FLAGS.GREAT_FAIRY,
          Settings.FLAGS.PUZZLE_SECRET_CAVE,
          Settings.FLAGS.FREE_GIFT,
          Settings.FLAGS.EXPENSIVE_PURCHASE,
          Settings.FLAGS.OTHER_CHEST,
          Settings.FLAGS.MISC,
        ]);
      });
    });

    describe('when using a development build', () => {
      beforeEach(() => {
        // version = 1.10.0_84fa5b4
        Settings.initializeFromPermalink('MS4xMC4wXzg0ZmE1YjQAeWVldAAHAQMADKC+AwDQAAIAAAAAAYAgAAA=');
      });

      test('sets the version to be the commit hash', () => {
        expect(Settings.version).toEqual('84fa5b4');
      });
    });

    describe('when using a beta build', () => {
      beforeEach(() => {
        // version = 1.10.0-BETA_2022-11-28
        Settings.initializeFromPermalink('MS4xMC4wLUJFVEFfMjAyMi0xMS0yOAB5ZWV0AAcBAwAMoL4DANAAAgAAAAABgCAAAA==');
      });

      test('sets the version to master', () => {
        expect(Settings.version).toEqual('master');
      });
    });

    describe('all flags set', () => {
      beforeEach(() => {
        Settings.initializeFromPermalink('MS4xMC4wAHllZXQA//8/AAygvgMA0AACAAAAAAGAIAAA');
      });

      test('initializes all the flags', () => {
        expect(_.difference(FLAGS, Settings.flags)).toEqual([]);
        expect(_.difference(Settings.flags, FLAGS)).toEqual([]);
      });
    });

    describe('all starting gear set', () => {
      beforeEach(() => {
        Settings.initializeFromPermalink('MS4xMC4wAHllZXQABwEDAAygvgMA0AD///////931QUA');
      });

      test('initializes the starting gear', () => {
        expect(Settings.startingGear).toMatchSnapshot();
      });
    });

    describe('only sunken triforce enabled', () => {
      describe('when charts are not randomized', () => {
        beforeEach(() => {
          Settings.initializeFromPermalink('MS4xMC4wAHllZXQAAEAAAAygvgMA0AACAAAAAAGAIAAA');
        });

        test('initializes the flags', () => {
          expect(Settings.flags).toEqual([
            Settings.FLAGS.SUNKEN_TRIFORCE,
          ]);
        });
      });

      describe('when charts are randomized', () => {
        beforeEach(() => {
          Settings.initializeFromPermalink('MS4xMC4wAHllZXQAAEAAgAygvgMA0AACAAAAAAGAIAAA');
        });

        test('initializes the flags', () => {
          expect(Settings.flags).toEqual([
            Settings.FLAGS.SUNKEN_TREASURE,
          ]);
        });
      });
    });
  });

  describe('initializeRaw', () => {
    test('initializes the flags, options, starting gear, and version', () => {
      const flags = [Settings.FLAGS.TINGLE_CHEST];
      const options = {
        [Permalink.OPTIONS.RANDOMIZE_ENTRANCES]: true,
      };
      const startingGear = {
        'Grappling Hook': 1,
      };
      const version = '1.0.0';

      Settings.initializeRaw({
        flags,
        options,
        startingGear,
        version,
      });

      expect(Settings.flags).toEqual(flags);
      expect(Settings.options).toEqual(options);
      expect(Settings.startingGear).toEqual(startingGear);
      expect(Settings.version).toEqual(version);
    });
  });

  describe('reset', () => {
    beforeEach(() => {
      Settings.initializeRaw({
        flags: [Settings.FLAGS.TINGLE_CHEST],
        options: {
          [Permalink.OPTIONS.RANDOMIZE_ENTRANCES]: true,
        },
        startingGear: {
          'Grappling Hook': 1,
        },
        version: '1.0.0',
      });
    });

    test('resets all the settings', () => {
      Settings.reset();

      expect(Settings.flags).toEqual(null);
      expect(Settings.options).toEqual(null);
      expect(Settings.startingGear).toEqual(null);
      expect(Settings.version).toEqual(null);
    });
  });

  describe('readAll', () => {
    let expectedFlags;
    let expectedOptions;
    let expectedStartingGear;
    let expectedVersion;

    beforeEach(() => {
      expectedFlags = [Settings.FLAGS.TINGLE_CHEST];
      expectedOptions = {
        [Permalink.OPTIONS.RANDOMIZE_ENTRANCES]: true,
      };
      expectedStartingGear = {
        'Grappling Hook': 1,
      };
      expectedVersion = '1.0.0';

      Settings.initializeRaw({
        flags: expectedFlags,
        options: expectedOptions,
        startingGear: expectedStartingGear,
        version: expectedVersion,
      });
    });

    test('returns all of the settings', () => {
      const allSettings = Settings.readAll();

      expect(allSettings).toEqual({
        flags: expectedFlags,
        options: expectedOptions,
        startingGear: expectedStartingGear,
        version: expectedVersion,
      });
    });
  });

  describe('FLAGS', () => {
    test('returns the correct flags', () => {
      expect(Settings.FLAGS).toMatchSnapshot();
    });
  });

  describe('isFlagActive', () => {
    beforeEach(() => {
      Settings.initializeRaw({
        flags: [Settings.FLAGS.TINGLE_CHEST],
      });
    });

    describe('when the flag is in the list of flags', () => {
      test('returns true', () => {
        const isFlagActive = Settings.isFlagActive(Settings.FLAGS.TINGLE_CHEST);

        expect(isFlagActive).toEqual(true);
      });
    });

    describe('when the flag is not in the list of flags', () => {
      test('returns true', () => {
        const isFlagActive = Settings.isFlagActive(Settings.FLAGS.DUNGEON);

        expect(isFlagActive).toEqual(false);
      });
    });
  });

  describe('getOptionValue', () => {
    beforeEach(() => {
      Settings.initializeRaw({
        options: {
          [Permalink.OPTIONS.NUM_STARTING_TRIFORCE_SHARDS]: 7,
        },
      });
    });

    test('returns the value of the given option', () => {
      const optionValue = Settings.getOptionValue(Permalink.OPTIONS.NUM_STARTING_TRIFORCE_SHARDS);

      expect(optionValue).toEqual(7);
    });
  });

  describe('getStartingGear', () => {
    beforeEach(() => {
      Settings.initializeRaw({
        startingGear: {
          'Deku Leaf': 1,
        },
      });
    });

    test('returns the provided starting gear', () => {
      const startingGear = Settings.getStartingGear();

      expect(startingGear).toEqual({ 'Deku Leaf': 1 });
    });
  });

  describe('getVersion', () => {
    beforeEach(() => {
      Settings.initializeRaw({
        version: '1.0.0',
      });
    });

    test('returns the provided version', () => {
      const version = Settings.getVersion();

      expect(version).toEqual('1.0.0');
    });
  });
});
