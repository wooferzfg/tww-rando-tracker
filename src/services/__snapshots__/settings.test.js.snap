// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settings FLAGS returns the correct flags 1`] = `
{
  "BATTLESQUID": "Battlesquid",
  "BIG_OCTO": "Big Octo",
  "COMBAT_SECRET_CAVE": "Combat Secret Cave",
  "DUNGEON": "Dungeon",
  "EXPENSIVE_PURCHASE": "Expensive Purchase",
  "EYE_REEF_CHEST": "Eye Reef Chest",
  "FREE_GIFT": "Free Gift",
  "GREAT_FAIRY": "Great Fairy",
  "GUNBOAT": "Gunboat",
  "ISLAND_PUZZLE": "Island Puzzle",
  "LONG_SIDEQUEST": "Long Sidequest",
  "MAIL": "Mail",
  "MINIGAME": "Minigame",
  "MISC": "Misc",
  "OTHER_CHEST": "Other Chest",
  "PLATFORM": "Platform",
  "PUZZLE_SECRET_CAVE": "Puzzle Secret Cave",
  "RAFT": "Raft",
  "SAVAGE_LABYRINTH": "Savage Labyrinth",
  "SHORT_SIDEQUEST": "Short Sidequest",
  "SPOILS_TRADING": "Spoils Trading",
  "SUBMARINE": "Submarine",
  "SUNKEN_TREASURE": "Sunken Treasure",
  "SUNKEN_TRIFORCE": "Sunken Triforce",
  "TINGLE_CHEST": "Tingle Chest",
}
`;

exports[`Settings initializeFromPermalink all starting gear set initializes the starting gear 1`] = `
{
  "Bait Bag": 1,
  "Ballad of Gales": 1,
  "Beedle's Chart": 1,
  "Bombs": 1,
  "Boomerang": 1,
  "Cabana Deed": 1,
  "Command Melody": 1,
  "DRC Compass": 1,
  "DRC Dungeon Map": 1,
  "Deku Leaf": 1,
  "Delivery Bag": 1,
  "Din's Pearl": 1,
  "ET Compass": 1,
  "ET Dungeon Map": 1,
  "Earth God's Lyric": 1,
  "Empty Bottle": 1,
  "FF Compass": 1,
  "FF Dungeon Map": 1,
  "FW Compass": 1,
  "FW Dungeon Map": 1,
  "Farore's Pearl": 1,
  "Fill-Up Coupon": 1,
  "Ghost Ship Chart": 1,
  "Grappling Hook": 1,
  "Great Fairy Chart": 1,
  "Hero's Charm": 1,
  "Hookshot": 1,
  "Hurricane Spin": 1,
  "Iron Boots": 1,
  "Light Ring Chart": 1,
  "Maggie's Letter": 1,
  "Magic Armor": 1,
  "Moblin's Letter": 1,
  "Nayru's Pearl": 1,
  "Note to Mom": 1,
  "Octo Chart": 1,
  "Platform Chart": 1,
  "Power Bracelets": 1,
  "Progressive Bomb Bag": 2,
  "Progressive Bow": 3,
  "Progressive Magic Meter": 2,
  "Progressive Picto Box": 2,
  "Progressive Quiver": 2,
  "Progressive Shield": 2,
  "Progressive Sword": 3,
  "Progressive Wallet": 2,
  "Secret Cave Chart": 1,
  "Skull Hammer": 1,
  "Song of Passing": 1,
  "Spoils Bag": 1,
  "Submarine Chart": 1,
  "Telescope": 1,
  "Tingle Tuner": 1,
  "Tingle's Chart": 1,
  "TotG Compass": 1,
  "TotG Dungeon Map": 1,
  "WT Compass": 1,
  "WT Dungeon Map": 1,
  "Wind God's Aria": 1,
}
`;

exports[`Settings initializeFromPermalink default settings initializes the options 1`] = `
{
  "add_shortcut_warps_between_dungeons": false,
  "chest_type_matches_contents": false,
  "cryptic_hints": true,
  "do_not_generate_spoiler_log": false,
  "fishmen_hints": true,
  "hoho_hints": true,
  "instant_text_boxes": true,
  "keylunacy": false,
  "korl_hints": false,
  "num_barren_hints": 0,
  "num_item_hints": 15,
  "num_location_hints": 5,
  "num_path_hints": 0,
  "num_race_mode_dungeons": 4,
  "num_starting_triforce_shards": 0,
  "prioritize_remote_hints": false,
  "progression_battlesquid": false,
  "progression_big_octos_gunboats": false,
  "progression_combat_secret_caves": false,
  "progression_dungeons": true,
  "progression_expensive_purchases": true,
  "progression_eye_reef_chests": false,
  "progression_free_gifts": true,
  "progression_great_fairies": true,
  "progression_island_puzzles": false,
  "progression_long_sidequests": false,
  "progression_mail": false,
  "progression_minigames": false,
  "progression_misc": true,
  "progression_platforms_rafts": false,
  "progression_puzzle_secret_caves": true,
  "progression_savage_labyrinth": false,
  "progression_short_sidequests": false,
  "progression_spoils_trading": false,
  "progression_submarines": false,
  "progression_tingle_chests": false,
  "progression_treasure_charts": false,
  "progression_triforce_charts": false,
  "race_mode": false,
  "randomize_charts": false,
  "randomize_enemies": false,
  "randomize_enemy_palettes": false,
  "randomize_entrances": "Disabled",
  "randomize_music": false,
  "randomize_starting_island": false,
  "remove_music": false,
  "reveal_full_sea_chart": true,
  "seed_name": "yeet",
  "skip_rematch_bosses": true,
  "starting_gear": {
    "Bait Bag": 0,
    "Ballad of Gales": 1,
    "Beedle's Chart": 0,
    "Bombs": 0,
    "Boomerang": 0,
    "Cabana Deed": 0,
    "Command Melody": 0,
    "DRC Compass": 0,
    "DRC Dungeon Map": 0,
    "Deku Leaf": 0,
    "Delivery Bag": 0,
    "Din's Pearl": 0,
    "ET Compass": 0,
    "ET Dungeon Map": 0,
    "Earth God's Lyric": 0,
    "Empty Bottle": 0,
    "FF Compass": 0,
    "FF Dungeon Map": 0,
    "FW Compass": 0,
    "FW Dungeon Map": 0,
    "Farore's Pearl": 0,
    "Fill-Up Coupon": 0,
    "Ghost Ship Chart": 0,
    "Grappling Hook": 0,
    "Great Fairy Chart": 0,
    "Hero's Charm": 0,
    "Hookshot": 0,
    "Hurricane Spin": 0,
    "Iron Boots": 0,
    "Light Ring Chart": 0,
    "Maggie's Letter": 0,
    "Magic Armor": 0,
    "Moblin's Letter": 0,
    "Nayru's Pearl": 0,
    "Note to Mom": 0,
    "Octo Chart": 0,
    "Platform Chart": 0,
    "Power Bracelets": 0,
    "Progressive Bomb Bag": 0,
    "Progressive Bow": 0,
    "Progressive Magic Meter": 1,
    "Progressive Picto Box": 0,
    "Progressive Quiver": 0,
    "Progressive Shield": 1,
    "Progressive Sword": 0,
    "Progressive Wallet": 0,
    "Secret Cave Chart": 0,
    "Skull Hammer": 0,
    "Song of Passing": 1,
    "Spoils Bag": 0,
    "Submarine Chart": 0,
    "Telescope": 0,
    "Tingle Tuner": 0,
    "Tingle's Chart": 0,
    "TotG Compass": 0,
    "TotG Dungeon Map": 0,
    "WT Compass": 0,
    "WT Dungeon Map": 0,
    "Wind God's Aria": 0,
  },
  "starting_hcs": 0,
  "starting_pohs": 0,
  "swift_sail": true,
  "sword_mode": "Start with Hero's Sword",
  "version": "1.10.0",
}
`;

exports[`Settings initializeFromPermalink default settings initializes the starting gear 1`] = `
{
  "Bait Bag": 0,
  "Ballad of Gales": 1,
  "Beedle's Chart": 0,
  "Bombs": 0,
  "Boomerang": 0,
  "Cabana Deed": 0,
  "Command Melody": 0,
  "DRC Compass": 0,
  "DRC Dungeon Map": 0,
  "Deku Leaf": 0,
  "Delivery Bag": 0,
  "Din's Pearl": 0,
  "ET Compass": 0,
  "ET Dungeon Map": 0,
  "Earth God's Lyric": 0,
  "Empty Bottle": 0,
  "FF Compass": 0,
  "FF Dungeon Map": 0,
  "FW Compass": 0,
  "FW Dungeon Map": 0,
  "Farore's Pearl": 0,
  "Fill-Up Coupon": 0,
  "Ghost Ship Chart": 0,
  "Grappling Hook": 0,
  "Great Fairy Chart": 0,
  "Hero's Charm": 0,
  "Hookshot": 0,
  "Hurricane Spin": 0,
  "Iron Boots": 0,
  "Light Ring Chart": 0,
  "Maggie's Letter": 0,
  "Magic Armor": 0,
  "Moblin's Letter": 0,
  "Nayru's Pearl": 0,
  "Note to Mom": 0,
  "Octo Chart": 0,
  "Platform Chart": 0,
  "Power Bracelets": 0,
  "Progressive Bomb Bag": 0,
  "Progressive Bow": 0,
  "Progressive Magic Meter": 1,
  "Progressive Picto Box": 0,
  "Progressive Quiver": 0,
  "Progressive Shield": 1,
  "Progressive Sword": 0,
  "Progressive Wallet": 0,
  "Secret Cave Chart": 0,
  "Skull Hammer": 0,
  "Song of Passing": 1,
  "Spoils Bag": 0,
  "Submarine Chart": 0,
  "Telescope": 0,
  "Tingle Tuner": 0,
  "Tingle's Chart": 0,
  "TotG Compass": 0,
  "TotG Dungeon Map": 0,
  "WT Compass": 0,
  "WT Dungeon Map": 0,
  "Wind God's Aria": 0,
}
`;
